src/littlefs.c

```diff
@@ -0,0 +1,110 @@
+#include "littlefs.h"
+#include "w25x.h"
+
+#define BLOCK_SIZE 4096
+#define BLOCK_COUNT 2048
+#define FLASH_SIZE (BLOCK_SIZE * BLOCK_COUNT)
+
+// 512 kB for dump (128 x 4kB)
+#define DUMP_BLOCK_COUNT 128
+
+// Address offset to skip area reserved for the dump
+#define ADDR_OFFSET (BLOCK_SIZE * DUMP_BLOCK_COUNT)
+
+lfs_t lfs;
+
+// Read a region in a block. Negative error codes are propogated
+// to the user.
+int read(const struct lfs_config *c, lfs_block_t block,
+        lfs_off_t off, void *buffer, lfs_size_t size);
+
+// Program a region in a block. The block must have previously
+// been erased. Negative error codes are propogated to the user.
+// May return LFS_ERR_CORRUPT if the block should be considered bad.
+int prog(const struct lfs_config *c, lfs_block_t block,
+        lfs_off_t off, const void *buffer, lfs_size_t size);
+
+// Erase a block. A block must be erased before being programmed.
+// The state of an erased block is undefined. Negative error codes
+// are propogated to the user.
+// May return LFS_ERR_CORRUPT if the block should be considered bad.
+int erase(const struct lfs_config *c, lfs_block_t block);
+
+// Sync the state of the underlying block device. Negative error codes
+// are propogated to the user.
+int sync(const struct lfs_config *c);
+
+// configuration of the filesystem is provided by this struct
+const struct lfs_config cfg = {
+    // block device operations
+    .read  = read,
+    .prog  = prog,
+    .erase = erase,
+    .sync  = sync,
+
+    // block device configuration
+    .read_size = 1,
+    .prog_size = 1,
+    .block_size = BLOCK_SIZE,
+    .block_count = BLOCK_COUNT - DUMP_BLOCK_COUNT,
+    .cache_size = 16,
+    .lookahead_size = 16,
+    .block_cycles = 500,
+};
+
+int read(const struct lfs_config *c, lfs_block_t block,
+        lfs_off_t off, void *buffer, lfs_size_t size) {
+
+    uint32_t addr = ADDR_OFFSET + block * BLOCK_SIZE + off;
+    if (addr >= FLASH_SIZE) {
+        return LFS_ERR_INVAL;
+    }
+
+    w25x_rd_data(addr, buffer, size);
+    return 0;
+}
+
+int prog(const struct lfs_config *c, lfs_block_t block,
+        lfs_off_t off, const void *buffer, lfs_size_t size) {
+
+    uint32_t addr = ADDR_OFFSET + block * BLOCK_SIZE + off;
+    if (addr >= FLASH_SIZE) {
+        return LFS_ERR_INVAL;
+    }
+
+    w25x_page_program(addr, (uint8_t*) buffer, size);
+    return 0;
+}
+
+int erase(const struct lfs_config *c, lfs_block_t block) {
+    uint32_t addr = ADDR_OFFSET + block * BLOCK_SIZE;
+    if (addr >= FLASH_SIZE) {
+        return LFS_ERR_INVAL;
+    }
+    w25x_sector_erase(addr);
+    return 0;
+}
+
+int sync(const struct lfs_config *c) {
+    w25x_wait_busy();
+    return 0;
+}
+
+lfs_t* littlefs_init() {
+    int err = lfs_mount(&lfs, &cfg);
+
+    // reformat if we can't mount the filesystem
+    // this should only happen on the first boot
+    if (err) {
```

Comment:
Personally, I'd allow the caller to choose what to do with the partition in case of an error. For debugging purposes, it may be beneficial to not format the XFLASH area but allow it to dump somewhere for investigation.
Please consider adding a parameter saying something like "format_on_error" ...

Reply:
This object is for initialization of the littlefs, meaning this is the caller that is deciding what to do. For debugging purpose this is the place where we can place a code with a different logic.